ode/unit_test/Test_ODE_BDF.hpp

```diff
@@ -0,0 +1,207 @@
+          Kokkos::pow(num_steps[idx - 1] / num_steps[idx], 2);
+      double actual_ratio = errors[idx] / errors[idx - 1];
+      std::cout << "expected ratio: " << expected_ratio
+                << ", actual ratio: " << actual_ratio << std::endl;
+    }
+  }
+
+  // Test BDF3
+  std::cout << "\nBDF3 convergence test" << std::endl;
+  for (int idx = 0; idx < num_tests; idx++) {
+    mv_type y_vecs("history vectors", mySys.neqs, 2);
+    Kokkos::deep_copy(y0, 0.5);
+
+    KokkosODE::Experimental::BDF<
+        KokkosODE::Experimental::BDF_type::BDF3>::Solve(mySys, t_start, t_end,
+                                                        num_steps[idx], y0,
+                                                        y_new, rhs, update,
+                                                        y_vecs, temp, jac);
+
+    errors[idx] = Kokkos::abs(y_new(0) - 1 / (1 + Kokkos::exp(-t_end))) /
+                  Kokkos::abs(1 / (1 + Kokkos::exp(-t_end)));
+    std::cout << y_new(0) << ", error=" << errors[idx] << std::endl;
+    if (idx > 0) {
+      double expected_ratio =
+          Kokkos::pow(num_steps[idx - 1] / num_steps[idx], 3);
+      double actual_ratio = errors[idx] / errors[idx - 1];
+      std::cout << "expected ratio: " << expected_ratio
+                << ", actual ratio: " << actual_ratio << std::endl;
+    }
+  }
+
+  // Test BDF4
+  std::cout << "\nBDF4 convergence test" << std::endl;
+  for (int idx = 0; idx < num_tests; idx++) {
+    mv_type y_vecs("history vectors", mySys.neqs, 2);
+    Kokkos::deep_copy(y0, 0.5);
+
+    KokkosODE::Experimental::BDF<
+        KokkosODE::Experimental::BDF_type::BDF4>::Solve(mySys, t_start, t_end,
+                                                        num_steps[idx], y0,
+                                                        y_new, rhs, update,
+                                                        y_vecs, temp, jac);
+
+    errors[idx] = Kokkos::abs(y_new(0) - 1 / (1 + Kokkos::exp(-t_end))) /
+                  Kokkos::abs(1 / (1 + Kokkos::exp(-t_end)));
+    std::cout << y_new(0) << ", error=" << errors[idx] << std::endl;
+    if (idx > 0) {
+      double expected_ratio =
+          Kokkos::pow(num_steps[idx - 1] / num_steps[idx], 4);
+      double actual_ratio = errors[idx] / errors[idx - 1];
+      std::cout << "expected ratio: " << expected_ratio
+                << ", actual ratio: " << actual_ratio << std::endl;
+    }
+  }
+
+  // Test BDF5
+  std::cout << "\nBDF5 convergence test" << std::endl;
+  for (int idx = 0; idx < num_tests; idx++) {
+    mv_type y_vecs("history vectors", mySys.neqs, 2);
+    Kokkos::deep_copy(y0, 0.5);
+
+    KokkosODE::Experimental::BDF<
+        KokkosODE::Experimental::BDF_type::BDF5>::Solve(mySys, t_start, t_end,
+                                                        num_steps[idx], y0,
+                                                        y_new, rhs, update,
+                                                        y_vecs, temp, jac);
```

Comment:
Eventually at the app level we're going to need to wrap things in an interface that is common between the explicit & implicit solvers. I don't think we need to worry about it in this PR, and maybe it makes more sense for that wrapper level to stay in app code, but I think we should at least consider having it in kokkos-kernels.

The interfaces are pretty much the same except for the different number of scratch views needed for extra space internally right now right?

Reply:
Correct, I am still waiting for Adams method before going with the final method which might be called `KokkosODE::Experimental::Solve()` and will allow to pick any specific method under the hood via template+params.
The internal storage will be lumped into a single view eventually but I still want Adams to get a good sense of all the scenario